File: internal/wraith/notify.go
package wraith

import (
	"fmt"
	"log"
	"os/exec"
	"strings"
)

// NotifyiMessage sends a message via macOS iMessage using AppleScript.
func NotifyiMessage(text, to string) error {
	if to == "" {
		to = "+18704160146" // General's phone
	}

	// Truncate for iMessage
	if len(text) > 800 {
		text = text[:800] + "..."
	}

	script := fmt.Sprintf(`tell application "Messages"
	set targetService to 1st account whose service type = iMessage
	set targetBuddy to participant "%s" of targetService
	send "%s" to targetBuddy
end tell`, to, escapeAppleScript(text))

	cmd := exec.Command("osascript", "-e", script)
	out, err := cmd.CombinedOutput()
	if err != nil {
		log.Printf("wraith/notify: iMessage failed: %s %v", string(out), err)
		return err
	}
	return nil
}

// NotifyIngestion sends a triage summary via iMessage.
func NotifyIngestion(source string, results []TriageResult) {
	if len(results) == 0 {
		return
	}

	var adapt, keep, discard []TriageResult
	for _, r := range results {
		switch r.Class {
		case "ADAPT":
			adapt = append(adapt, r)
		case "KEEP":
			keep = append(keep, r)
		case "DISCARD":
			discard = append(discard, r)
		}
	}

	var sb strings.Builder
	fmt.Fprintf(&sb, "WRAITH: %d %s items ingested\n\n", len(results), source)

	if len(adapt) > 0 {
		fmt.Fprintf(&sb, "ADAPT (%d):\n", len(adapt))
		for _, r := range adapt {
			fmt.Fprintf(&sb, "  - %s\n    %s\n", truncate(r.Title, 60), truncate(r.Reason, 80))
		}
		sb.WriteString("\n")
	}
	if len(keep) > 0 {
		fmt.Fprintf(&sb, "KEEP (%d):\n", len(keep))
		for _, r := range keep[:min(5, len(keep))] {
			fmt.Fprintf(&sb, "  - %s\n", truncate(r.Title, 60))
		}
		sb.WriteString("\n")
	}
	if len(discard) > 0 {
		fmt.Fprintf(&sb, "DISCARD (%d)\n", len(discard))
	}

	if err := NotifyiMessage(sb.String(), ""); err != nil {
		log.Printf("wraith/notify: failed to send summary: %v", err)
	}
}

func escapeAppleScript(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, "\"", "\\\"")
	s = strings.ReplaceAll(s, "\n", "\\n")
	return s
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: wraith/notify.py
"""WRAITH ingestion notifications.

Sends brief iMessage summaries to the General after each ingestion run.
Uses AppleScript for delivery — same approach as services.imessage.daemon.
"""

from __future__ import annotations

import logging
import subprocess

log = logging.getLogger(__name__)

GENERAL_PHONE = "+18704160146"


def _send_imessage(text: str, to: str = GENERAL_PHONE) -> bool:
    """Send an iMessage via AppleScript. Returns True on success."""
    escaped = text.replace("\\", "\\\\").replace('"', '\\"')
    script = f'''
    tell application "Messages"
        set targetService to 1st account whose service type = iMessage
        set targetBuddy to participant "{to}" of targetService
        send "{escaped}" to targetBuddy
    end tell
    '''
    try:
        subprocess.run(
            ["/usr/bin/osascript", "-e", script],
            check=True, timeout=10, capture_output=True,
        )
        log.info("iMessage sent: %s", text[:80])
        return True
    except subprocess.TimeoutExpired:
        log.warning("iMessage send timed out")
        return False
    except Exception as exc:
        log.error("iMessage send failed: %s", exc)
        return False


def notify_ingestion(
    source: str,
    count: int,
    mode: str,
    warnings: list[str] | None = None,
) -> bool:
    """Send an iMessage notification summarizing a WRAITH ingestion run.

    Examples:
        WRAITH: 23 X bookmarks ingested (ghost mode)
        WRAITH: 0 items — Reddit cookies may be expired
    """
    if count > 0:
        msg = f"WRAITH: {count} {source} ingested ({mode} mode)"
    else:
        msg = f"WRAITH: 0 {source} items"

    if warnings:
        msg += " — " + "; ".join(warnings)

    return _send_imessage(msg)
